GVRf/Framework/src/org/gearvrf/io/GVRInputManager.java

```diff
@@ -149,36 +149,46 @@ private GVRCursorType getGVRInputDeviceType(InputDevice device) {
                             : GVRCursorType.UNKNOWN);
                 }
                 return GVRCursorType.MOUSE;
-            } else if ((device.getSources()
-                    & InputDevice.SOURCE_GAMEPAD) == InputDevice.SOURCE_GAMEPAD
-                    || (device.getSources()
-                            & InputDevice.SOURCE_JOYSTICK) == InputDevice.SOURCE_JOYSTICK) {
+            } else if ((sources & InputDevice.SOURCE_GAMEPAD) == InputDevice.SOURCE_GAMEPAD
+                    || (sources & InputDevice.SOURCE_JOYSTICK) == InputDevice.SOURCE_JOYSTICK) {
                 return GVRCursorType.CONTROLLER;
+            } else if ((sources & InputDevice.SOURCE_KEYBOARD) == InputDevice.SOURCE_KEYBOARD) {
+                // Allow device keyboard to be a gaze controller if enabled
+                return useGazeCursorController
+                        ? GVRCursorType.GAZE
+                        : GVRCursorType.UNKNOWN;
             }
         }
         return GVRCursorType.UNKNOWN;
     }
 
-    // Return the key if there is one else return -1
-    private long getCacheKey(InputDevice device, GVRCursorType cursorType) {
+    // Return the key if there is one else return 0 (note: -1 is the ID of an Android input device)
+    private long getCacheKey(InputDevice device, int deviceId, GVRCursorType cursorType) {
         if (cursorType != GVRCursorType.UNKNOWN
                 && cursorType != GVRCursorType.EXTERNAL) {
             // Sometimes a device shows up using two device ids
             // here we try to show both devices as one using the
             // product and vendor id
-            long key = device.getVendorId() << 32;
-            key = key | device.getProductId();
+            long key = 0;
+            if (device.getProductId() != 0 && device.getVendorId() != 0) {
+                key = device.getVendorId() << 32;
+                key = key | device.getProductId();
+            } else {
+                // If there is no product/vendor ID (non-USB device), use device Id.
+                // The key/deviceId can be -1 for android's default input
+                key = deviceId;
```

Comment:
Yeah. I've seen multiple gaze controllers being created too. It would be nice for them to share the same one. What's the best way of doing that? If you could help, that'll be great too.